### DISCUSSION TITLE
Why does sqlalchemy's default sqlite datetime format have 6 digits of precision for microseconds?

### DISCUSSION THREAD

**Discussion by lukesmurray** [sqlalchemy/sqlalchemy#11635] [Usage Questions] [answered]
SQLITE date functions only support 3 digits of precision after the decimal point [link](https://www.sqlite.org/lang_datefunc.html#:~:text=or%20unixepoch()%20calls.-,2.%20Time%20Values,-A%20time%2Dvalue)

> In formats 4, 7, and 10, the fractional seconds value SS.SSS can have one or more digits following the decimal point. Exactly three digits are shown in the examples because only the first three digits are significant to the result, but the input string can have fewer or more than three digits and the date/time functions will still operate correctly. Similarly, format 12 is shown with 10 significant digits, but the date/time functions will really accept as many or as few digits as are necessary to represent the Julian day number.

I'm curious why the default sqlite datetime format uses 6 digits of precision by default after the decimal point?

https://docs.sqlalchemy.org/en/20/dialects/sqlite.html#sqlalchemy.dialects.sqlite.DATETIME

I believe this can lead to subtle bugs. For example if I use `strftime` to create a default `created_at` timestamp - the most precision I can use with built in sqlite functions is 3 digits after the decimal point.

This led to a bug when I was using cursor based pagination and tried to select rows that were less than a given `created_at` date. Because sqlite doesn't have a date type this becomes a `TEXT` comparison and the mismatch in precision matters.

Here's an example where I compare two strings that are identical except for additional 0s after the decimal point on the right.

```sql
select
'2013-10-07 08:23:19.120' < '2013-10-07 08:23:19.120000';
-- TRUE
```

**Comment by CaselIT** [comment]
Hi,

From the blame it seems that it has always been like that. I'm guessing it's to round trip pyrhon date times that support 6 digits of microsecond precision

**Comment by zzzeek** [comment, marked as answer]
The SQlite3 adapter itself includes a default adapter that [also uses a 6 digit microsecond format](https://docs.python.org/3/library/sqlite3.html#default-adapters-and-converters-deprecated) (which is also maintained in their new [recipe](https://docs.python.org/3/library/sqlite3.html#sqlite3-adapter-converter-recipes) version).    Python datetime objects support this level of precision so first and foremost that's what people want to be able to round-trip from SQLite, which has no datetime type of its own.   This is the standard approach to persisting Python datetimes.

The date information I see linked at sqlite's documentation might have been written after SQLAlchemy's first releases in 2006, so I dont really know the origin of their date functions.    Since the document stresses that their current millisecond format is nothing more than ["the current implementation", which "might increase to a higher resolution in future releases of SQLite"](https://www.sqlite.org/lang_datefunc.html#modifiers) I would reach out to them and see if there are plans to increase this to 6 digits.

**Reply by lukesmurray** [reply]
thanks for the thoughtful reply.

**Comment by lukesmurray** [comment]
another solution to this is to use `STRFTIME` with additional precision. Then there is no need to override any of sqlalchemy's defaults.

```
STRFTIME('%Y-%m-%d %H:%M:%f000', 'NOW')
```
